# **Pandas Basics and DataFrame Manipulation**

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# 📖 TABLE OF CONTENTS

- [Setup](#Setup)
- [Loading and Viewing Data in Pandas](#Loading-and-Viewing-Data-in-Pandas)
  - [Installing and Importing pandas](#Installing-and-Importing-pandas)
  - [Introduction to DataFrames](#Introduction-to-DataFrames)
  - [Creating from Dictionary](#Creating-from-Dictionary)
  - [Viewing Data in a DataFrame: Head and Tail](#Viewing-Data-in-a-DataFrame:-Head-and-Tail)
  - [Viewing Data in a DataFrame: Overview](#Viewing-Data-in-a-DataFrame:-Overview)
  - [Concatenating DataFrames: `pd.concat`](#Concatenating-DataFrames:-`pd.concat`)
  - [Series](#Series)
  - [Lesson Summary](#Lesson-Summary)
- [Understanding Your Data](#Understanding-Your-Data)
  - [Basic Statistical Quantities](#Basic-Statistical-Quantities)
  - [Calculation of Statistical Quantities in Pandas](#Calculation-of-Statistical-Quantities-in-Pandas)
  - [DataFrame Describe Function](#DataFrame-Describe-Function)
  - [Lesson Summary](#Lesson-Summary)
- [Indexing and Selecting Data in Pandas](#Indexing-and-Selecting-Data-in-Pandas)
  - [Understanding Indexing: Setting Index](#Understanding-Indexing:-Setting-Index)
  - [Understanding Indexing: Resetting Index](#Understanding-Indexing:-Resetting-Index)
  - [Understanding Indexing: Renaming Index](#Understanding-Indexing:-Renaming-Index)
  - [Selecting Data using Labels and Location](#Selecting-Data-using-Labels-and-Location)
  - [Lesson Summary](#Lesson-Summary)
- [Filtering Data in Pandas](#Filtering-Data-in-Pandas)
  - [Basics of Data Filtering](#Basics-of-Data-Filtering)
  - [Understanding Boolean Masking](#Understanding-Boolean-Masking)
  - [Advanced Data Filtering](#Advanced-Data-Filtering)
  - [Lesson Summary](#Lesson-Summary)
- [Basic Data Analysis](#Basic-Data-Analysis)
  - [Meaning of Basic Data Analysis](#Meaning-of-Basic-Data-Analysis)
  - [Using `value_counts()` for Frequency Analysis](#Using-`value_counts()`-for-Frequency-Analysis)
  - [Grouping and Aggregating with `groupby()` and `agg()` methods](#Grouping-and-Aggregating-with-`groupby()`-and-`agg()`-methods)
  - [Sorting DataFrame with `sort_values()`](#Sorting-DataFrame-with-`sort_values()`)
  - [Lesson Summary](#Lesson-Summary)
- [References](#References)

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# Setup

Before we jump in, let's quickly run the setup cell. It checks the environment (GPU, library versions) and sets random seeds for reproducibility — we want identical results on every run.

In [1]:
# 🔧 Setup: Run this cell first!
# Environment check + random seeds for reproducibility

import sys
import time
import random
import numpy as np
import pandas as pd
import matplotlib as mpl
import torch
import torch.nn as nn

print(f"📦 Python  {sys.version.split()[0]}")
print(f"🔢 NumPy   {np.__version__}")
print(f"📊 Matplotlib   {mpl.__version__}")
print(f"🔥 PyTorch {torch.__version__}")

# GPU check (only needed for optional deep-feature experiments)
if torch.cuda.is_available():
    device = torch.device("cuda")
    print(f"✅ GPU available: {torch.cuda.get_device_name(0)}")
    print(f"   Memory: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")
else:
    device = torch.device("cpu")
    print("⚠️ No GPU detected")

DEVICE_LABEL = "GPU" if device.type == "cuda" else "CPU"

# Random seeds for reproducibility
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed_all(SEED)

print(f"\n🎲 Random seed set to {SEED} (Python, NumPy, PyTorch)")

%matplotlib inline
%config InlineBackend.figure_format = 'jpeg'   # photos as JPEG, not PNG — ~10× smaller notebook
# Remove character limits for all columns of Pandas dataframes permanently in this session
pd.set_option('display.max_colwidth', None)

📦 Python  3.13.15
🔢 NumPy   2.1.3
📊 Matplotlib   3.10.0
🔥 PyTorch 2.11.0+cpu
⚠️ No GPU detected

🎲 Random seed set to 42 (Python, NumPy, PyTorch)


![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# Loading and Viewing Data in Pandas

Hello and welcome to our journey into data analysis with Python and **pandas**. Today we'll discover **pandas DataFrames** and learn about *Loading and Viewing Data*.

**Pandas**, a fantastic Python library, simplifies data manipulation and analysis. Our focus today is `DataFrames` — the go-to structure in pandas for data handling.

We will read data from different sources using `pandas`, load it into a `DataFrame`, and then explore this data. Let's begin!

## Installing and Importing pandas

In [ ]:
import pandas as pd

This line sets a short alias, `pd`, for `pandas` so we don't have to write out `pandas` each time we use it.

## Introduction to DataFrames

In `pandas`, a `DataFrame` is like a table, with the data as the dishes on the table. Creating a `DataFrame` out of a list or a dictionary is a snap with `pandas`. Here's how:

In [ ]:
import pandas as pd

# From a list
data_list = ['apple', 'banana', 'cherry']
df_list = pd.DataFrame(data_list, columns=['Fruit'])
print(df_list)

    Fruit
0   apple
1  banana
2  cherry


## Creating from Dictionary

And here is how to create a dataframe from dictionary:

In [ ]:
# From a dictionary
data_dict = {'Fruit': ['apple', 'banana', 'cherry'], 'Count': [10, 20, 15]}
df_dict = pd.DataFrame(data_dict)
print(df_dict)

    Fruit  Count
0   apple     10
1  banana     20
2  cherry     15


## Viewing Data in a DataFrame: Head and Tail

Now that we have our data in a `DataFrame`, how do we look at it and understand it? `Pandas` provides us with methods like `head()`, `tail()`, and `info()`. Here's how to use them:

In [ ]:
print(df_dict.head())  # Output: First 5 rows of DataFrame 'df_dict'
print('*' * 20)  # Separator
print(df_dict.tail())  # Output: Last 5 rows of DataFrame 'df_dict'

    Fruit  Count
0   apple     10
1  banana     20
2  cherry     15
********************
    Fruit  Count
0   apple     10
1  banana     20
2  cherry     15


In our case, we have just three rows in the dataframe, so both `head()` and `tail()` will simply output the whole dataframe. However, for real data with lots of rows, they are quite useful!

## Viewing Data in a DataFrame: Overview

Let's take a look at the dataframe's Overview:

In [ ]:
print(df_dict.info())  # Dataframe's overview

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fruit   3 non-null      object
 1   Count   3 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 180.0+ bytes
None


As you see, the overview contains information about column's names, amount of present data and data types for each column.

## Concatenating DataFrames: `pd.concat`

Sometimes, you might need to combine multiple DataFrames into a single one. This can be done using the `pd.concat` function for any dataframes with the same set of columns. Here's a simple example:

In [ ]:
# Creating first DataFrame
data1 = {'Fruit': ['apple', 'banana'], 'Count': [10, 20]}
df1 = pd.DataFrame(data1)

# Creating second DataFrame
data2 = {'Fruit': ['cherry', 'date'], 'Count': [15, 25]}
df2 = pd.DataFrame(data2)

# Concatenating DataFrames
df_combined = pd.concat([df1, df2])
print(df_combined)

    Fruit  Count
0   apple     10
1  banana     20
0  cherry     15
1    date     25


In this example, `pd.concat` takes a list of DataFrames as its argument and combines them along their rows by default. Notice that the indices are preserved. If you want to ignore the original indices and create a new continuous index, you can pass the argument `ignore_index=True` to `pd.concat`:

In [ ]:
df_combined = pd.concat([df1, df2], ignore_index=True)
print(df_combined)

    Fruit  Count
0   apple     10
1  banana     20
2  cherry     15
3    date     25


This way, the resulting DataFrame will have a new set of sequential indices.

## Series

In a `DataFrame`, each column is a `Series` object. A `Series` in pandas is a one-dimensional labeled array capable of holding any data type (integers, strings, floating-point numbers, Python objects, etc.). It is essentially a list of values with an associated label (index) for each value. Here is an example:

In [ ]:
import pandas as pd

# Create a Series from a list
fruits = ['apple', 'banana', 'cherry']
series = pd.Series(fruits)
print(series)

0     apple
1    banana
2    cherry
dtype: object


Whenever you work with a single dataframe's column, you work with a series. Series objects have their own set of methods, but most of them overlap with the dataframe's method. For example, series also have methods like `head`, `tail` or `describe`.

## Lesson Summary

In this lesson, we've covered what `pandas` and a `DataFrame` are, how to load data into a `DataFrame`, and methods to view the data. It's a solid start to data analysis.

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# Understanding Your Data

Understanding your data is like plotting key positions on a map for a journey. How do we do it? Through statistical quantities like *mean*, *median*, and *mode*. These tell us more about our data. Today, we will learn about these quantities and how to calculate them in `pandas`.

## Basic Statistical Quantities

In data analysis, *mean*, *median*, and *mode* help us understand the central tendency of our data. The mean is the average, the median is the middle value when data is sorted in order, and the mode is the most frequent value. Standard deviation and variance tell us how data varies and the difference between each quantity and the mean. Additionally, we use minimum (`min`), maximum (`max`), and quantiles to understand data spread.

As a reminder, quantiles divide data into equal-sized intervals and help understand its distribution. One example of quantile is quartiles, which divide data into 4 equally sized groups. For instance, the first quartile is the value which is greater than the 25% of the data.

## Calculation of Statistical Quantities in Pandas

We compute these quantities for a DataFrame or Series object `data` using:

* mean = `data.mean()`,
* median = `data.median()`,
* mode = `data.mode()`,
* standard deviation = `data.std()`,
* variance = `data.var()`,
* min = `data.min()`,
* max = `data.max()`,
* quantile = `data.quantile(q)`, where `q` is the quantile like 0.25, 0.5, etc.

Let's calculate these for a sample DataFrame:

In [ ]:
import pandas as pd

# DataFrame creation
data = pd.DataFrame({
    'friends': ['Alice', 'Bob', 'Charlie', 'Dave', 'Eve'],
    'scores': [93, 89, 82, 88, 94],
    'age': [20, 21, 20, 19, 21]
})

# Print statistics
print(f"Mean: {data['scores'].mean()}")
print(f"Median: {data['scores'].median()}")
print(f"Mode: {data['scores'].mode()[0]}")
print(f"Standard Deviation: {data['scores'].std()}")
print(f"Variance: {data['scores'].var()}")
print(f"Min: {data['scores'].min()}")
print(f"Max: {data['scores'].max()}")
print(f"25% Quantile: {data['scores'].quantile(0.25)}")

Mean: 89.2
Median: 89.0
Mode: 82
Standard Deviation: 4.764451699828639
Variance: 22.700000000000003
Min: 82
Max: 94
25% Quantile: 88.0


## DataFrame Describe Function

`Pandas` provides `describe()`, which computes these statistical quantities for each DataFrame column. Here's how it looks:

In [ ]:
data.describe()

,scores,age
count,5.000000,5.00000
mean,89.200000,20.20000
std,4.764452,0.83666
min,82.000000,19.00000
25%,88.000000,20.00000
50%,89.000000,20.00000
75%,93.000000,21.00000
max,94.000000,21.00000


## Lesson Summary

Today, you learned about *mean*, *median*, *mode*, *standard deviation*, *variance*, *min*, *max*, and *quantiles*, and how these are calculated using `pandas`. You saw `describe()` being used on real-world data.

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# Indexing and Selecting Data in Pandas

Today we're diving into *Indexing and Selecting Data* in `pandas`, a crucial part of data manipulation and analysis. Indexing helps us locate data in specific rows while selecting focuses on picking specific columns or cells.

## Understanding Indexing: Setting Index

In `pandas`, an *index* is more or less the address of your data. By default, `pandas` assigns integer labels to the rows, but we can set any column as the index. This effectively turns it into an identifier for the rows.

Here's a basic example using `pandas` DataFrame's `set_index()` method:

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'John'],
    'Age': [25, 22, 30],
    'City': ['New York', 'Los Angeles', 'Chicago']
})

df.set_index('Name', inplace=True)
df

,Age,City
Name,,
Alice,25,New York
Bob,22,Los Angeles
John,30,Chicago


Accessing data using the index is performed with pandas `loc[]` method for label-based indexing and `iloc[]` method for integer-based indexing, which we will investigate later.

The `inplace` parameter is common for a lot of pandas dataframe methods. If `inplace` is set to `True`, changes are applied to the target dataframe. Otherwise, the target dataframe will be copied, the copy will be changed and returned.

However, it is important to note that in the `pandas 3.0` the `inplace` parameter will be omitted, and you will have to do it this way:

```python
df.set_index('Name')
```

## Understanding Indexing: Resetting Index

If you want to reset index back to the default, it is done easily with the following method:

In [ ]:
df.reset_index(inplace=True)
df

,Name,Age,City
0,Alice,25,New York
1,Bob,22,Los Angeles
2,John,30,Chicago


## Understanding Indexing: Renaming Index

Renaming the index is simply renaming the corresponding column. It is done with the `rename` method:

In [ ]:
df.rename(columns={'Name': 'Student Name', 'Age': 'Student Age'}, inplace=True)
df

,Student Name,Student Age,City
0,Alice,25,New York
1,Bob,22,Los Angeles
2,John,30,Chicago


Here, we provide a dictionary where the key is the old name, and the value is the new name.

## Selecting Data using Labels and Location

`Pandas` provides `loc[]` and `iloc[]` for accessing data in a DataFrame in a manner similar to array indexing for label-based and integer-based indexing, respectively. `loc[]` uses label-based indexing, and `iloc[]` uses integer-based indexing.

Let's understand this with an example:

In [ ]:
df = pd.DataFrame({
    'Name': ['Alice', 'Bob', 'John', 'Robert', 'Ann'],
    'Age': [25, 22, 30, 28, 32],
    'City': ['New York', 'Los Angeles', 'Chicago', 'San Francisco', 'Houston']
})

df.set_index('Name', inplace=True)

df.loc[['Alice', 'John'], ['Age', 'City']]

,Age,City
Name,,
Alice,25,New York
John,30,Chicago


In [ ]:
df.iloc[[1, 3], [0, 1]]

,Age,City
Name,,
Bob,22,Los Angeles
Robert,28,San Francisco


Note that we set the `'Name'` column as index. In `loc`, we use labels (which Is the name-indices and column names) to select the required data. In `iloc`, we use numerical indices for both rows and columns: It works similarly to 2d NumPy arrays.

In [ ]:
import pandas as pd

schools = pd.DataFrame({
    'Institution': ['Greenwood Primary', 'Sunshine Secondary', 'Harbor Science'],
    'Students': [320, 200, 250],
    'City': ['Cambridge', 'Oxford', 'Brighton']
})

schools.set_index('Institution', inplace=True)

schools.loc[['Greenwood Primary', 'Harbor Science'], 'City']

,City
Institution,
Greenwood Primary,Cambridge
Harbor Science,Brighton


In [ ]:
schools.loc[['Greenwood Primary', 'Harbor Science'], ['City']]

,City
Institution,
Greenwood Primary,Cambridge
Harbor Science,Brighton


In [ ]:
schools.iloc[[0, 2], 1]

,City
Institution,
Greenwood Primary,Cambridge
Harbor Science,Brighton


In [ ]:
schools.iloc[[0, 2], [1]]

,City
Institution,
Greenwood Primary,Cambridge
Harbor Science,Brighton


## Lesson Summary

You've learned how to index and select data in `pandas`, including functions like `set_index()`, `reset_index()`, `loc[]`, and `iloc[]`.

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# Filtering Data in Pandas

**Filtering Data** is all about focusing on the data that matters to us. By the end, you'll be able to pick the necessary data from a big data set using `pandas`.

## Basics of Data Filtering

Filtering data in `pandas` is like finding your favorite outfit in a wardrobe. The easiest way to filter data is by columns. Let's illustrate this using a `DataFrame` of students' details.

In [17]:
import pandas as pd

# Data of students
data = {
    'name': ['Alice', 'Bob', 'Charlie', 'Dave', 'Eve'],
    'age': [12, 13, 14, 13, 12],
    'grade_level': [6, 7, 8, 7, 6]
}

students_df = pd.DataFrame(data)

# Filter 7th grade students
grade_seven_students = students_df[students_df['grade_level'] == 7]

grade_seven_students

,name,age,grade_level
1,Bob,13,7
3,Dave,13,7


The code above creates a `DataFrame` and selects only the rows where the `grade_level` is 7. Now, you have the data of the 7th-grade students. Note that it works exactly the same as the numpy's boolean selection. Let's recall how it works under the hood.

## Understanding Boolean Masking

One of the magic tricks of `pandas` is Boolean masking. `Boolean` is a True or False data type. "Mask" means to hide. A Boolean mask hides parts of your data based on it being True or False.

We can create a Boolean Series, a list of True or False values, in `pandas` and use it for filtering.

In [18]:
is_grade_seven = students_df['grade_level'] == 7
is_grade_seven

,grade_level
0,False
1,True
2,False
3,True
4,False


This code creates a Boolean Series checking where the `grade_level` is 7. Then, it filters the data using this series:

In [19]:
# Filtering using Boolean series
grade_seven_students = students_df[is_grade_seven]
grade_seven_students

,name,age,grade_level
1,Bob,13,7
3,Dave,13,7


Note that only students with `True` in the boolean series were selected.

## Advanced Data Filtering

Sometimes we need to filter data using multiple conditions. Python lets us do this with logical operators: And (`&`), Or (`|`), and Not (`~`). Let's check them out:

In [21]:
# Filter 7th grade students who are 13 years old
grade_seven_and_age_thirteen = (students_df['grade_level'] == 7) & (students_df['age'] == 13)

students_df[grade_seven_and_age_thirteen]

,name,age,grade_level
1,Bob,13,7
3,Dave,13,7


**Note** Avoiding parantheses between logical operators (`&`, `|`, `~`) generates error.

The `isin()` method in `pandas` is another wonderful tool. It checks whether a `pandas` Series is in a list of values.

In [23]:
# Filter students who are in 6th or 7th grade
middle_school_students = students_df[students_df['grade_level'].isin([6, 7])]
middle_school_students

,name,age,grade_level
0,Alice,12,6
1,Bob,13,7
3,Dave,13,7
4,Eve,12,6


Fantastic! Now you know advanced data filtering techniques.

## Lesson Summary

This lesson covered basic to advanced data filtering, including Boolean masking and multiple conditions in filtering.

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# Basic Data Analysis

Today, we dive into **Basic Data Analysis** — an essential step for data comprehension. We unearth patterns, and guide decision-making across various fields, be it business, science, or daily life, with a powerful tool—the `pandas` Python library. Let's embark on this journey!

## Meaning of Basic Data Analysis

Rising to the challenge of solving a data mystery, **Basic Data Analysis** serves as the groundwork. It encompasses understanding and decision-making — be it a business owner understanding customer behavior, a scientist analyzing research data, or a student making sense of study material. With `pandas`, this process becomes effortless.

## Using `value_counts()` for Frequency Analysis

Firstly, we employ `value_counts()`, a method that swiftly counts the frequency of `DataFrame` elements. Consider an imaginary dataset of pets.

In [2]:
import pandas as pd

data = {
    'Name': ['Tommy', 'Rex', 'Bella', 'Charlie', 'Lucy', 'Cooper'],
    'Type': ['Dog', 'Dog', 'Cat', 'Cat', 'Dog', 'Bird']
}

pets_df = pd.DataFrame(data)
pets_df

,Name,Type
0,Tommy,Dog
1,Rex,Dog
2,Bella,Cat
3,Charlie,Cat
4,Lucy,Dog
5,Cooper,Bird


Using the `value_counts()` function we can count unique elements of a Pandas `Series` (a dataframe column):

In [3]:
pets_df['Type'].value_counts()

,count
Type,
Dog,3
Cat,2
Bird,1


With `value_counts()`, establishing frequency distribution in series becomes straightforward.

## Grouping and Aggregating with `groupby()` and `agg()` methods

For summarizing data, `groupby()` and `agg()` prove useful! Now let’s add weight to the pets in our DataFrame to illustrate these methods:

In [4]:
import pandas as pd

data = {
    'Name': ['Tommy', 'Rex', 'Bella', 'Charlie', 'Lucy', 'Cooper'],
    'Type': ['Dog', 'Dog', 'Cat', 'Cat', 'Dog', 'Bird'],
    'Weight': [12, 15, 8, 9, 14, 1]
}

pets_df = pd.DataFrame(data)
pets_df

,Name,Type,Weight
0,Tommy,Dog,12
1,Rex,Dog,15
2,Bella,Cat,8
3,Charlie,Cat,9
4,Lucy,Dog,14
5,Cooper,Bird,1


Now group and calculate the mean of data based on pet type.

In [5]:
pets_df.groupby('Type').agg({'Weight': 'mean'})

,Weight
Type,
Bird,1.000000
Cat,8.500000
Dog,13.666667


1. `groupby('Type')`: Splits the data into groups based on `'Type'`.
2. `.agg({'Weight': 'mean'})`: Applies the `'mean'` function to the `'Weight'` column for each group.

The resulting DataFrame shows the average weight for each pet type:
* Bird: 1.0
* Cat: 8.5
* Dog: 13.67

Of course, calculating `mean` is not the only option. We can use functions like `min`, `max`, `median`, etc. We will talk more about using different aggregation functions in the next course.

## Sorting DataFrame with `sort_values()`

Lastly, let's sort our data. The `sort_values()` function sorts our `DataFrame` as per one or many columns.

Let's arrange our pet `DataFrame` by pet weight.

In [6]:
sorted_pets_df = pets_df.sort_values('Weight')
sorted_pets_df

,Name,Type,Weight
5,Cooper,Bird,1
2,Bella,Cat,8
3,Charlie,Cat,9
0,Tommy,Dog,12
4,Lucy,Dog,14
1,Rex,Dog,15


In [7]:
descending_sorted_pets_df = pets_df.sort_values('Weight', ascending=False)
descending_sorted_pets_df

,Name,Type,Weight
1,Rex,Dog,15
4,Lucy,Dog,14
0,Tommy,Dog,12
3,Charlie,Cat,9
2,Bella,Cat,8
5,Cooper,Bird,1


We obtained sorted data efficiently with just one simple command!

## Lesson Summary

Great job! You've learned how to execute **Basic Data Analysis** using `pandas` functions. We explored `value_counts()`, `groupby()`, `agg()`, and `sort_values()`.

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)

# References

- **[Pandas Basics and DataFrame Manipulation](https://codesignal.com/learn/courses/pandas-basics-and-dataframe-manipulation)**

![rainbow](https://raw.githubusercontent.com/ancilcleetus/AI-Learning-Lab/main/assets/rainbow-divider.png)